src/include/storage/index/index_builder.h

```diff
@@ -78,33 +78,79 @@ class IndexBuilder {
                    "sql_table / txn is null and txn / sql_table is not.");
     sql_table_ = sql_table;
     txn_ = txn;
     return *this;
   }
-
- private:
  /**
   *
   * @param newly created index
   * @return index with all the keys inserted
   */
   void BulkInsert(Index *index) const {
-    uint32_t pr_size = index->GetProjectedRowInitializer().ProjectedRowSize();
-    byte *index_pr_buffer = common::AllocationUtil::AllocateAligned(pr_size);
+    const auto index_pr_initializer = index->GetProjectedRowInitializer();
+
+    const uint32_t index_pr_size = index_pr_initializer.ProjectedRowSize();
+    byte *index_pr_buffer = common::AllocationUtil::AllocateAligned(index_pr_size);
+
+    ProjectedRow *index_pr = index_pr_initializer.InitializeRow(index_pr_buffer);
+
+    const auto &indexed_attributes = key_schema_.GetIndexedColOids();
+    const auto table_pr_initializer = sql_table_->InitializerForProjectedRow(indexed_attributes);
 
-    ProjectedRow *index_pr = index->GetProjectedRowInitializer().InitializeRow(index_pr_buffer);
+    const uint32_t table_pr_size = table_pr_initializer.ProjectedRowSize();
+    byte *table_pr_buffer = common::AllocationUtil::AllocateAligned(table_pr_size);
+
+    ProjectedRow *table_pr = table_pr_initializer.InitializeRow(table_pr_buffer);
+
+    auto pr_map = sql_table_->ProjectionMapForOids(indexed_attributes);
 
     for (auto it = sql_table_->begin(); it != sql_table_->end(); ++it) {
       const TupleSlot slot = *it;
-      index->Insert(txn_, *index_pr, slot);
+
+      sql_table_->TraverseVersionChain(slot, table_pr, [&, this](auto type){
+        // Insert into the index
+        auto num_index_cols = key_schema_.GetColumns().size();
+        TERRIER_ASSERT(num_index_cols == indexed_attributes.size(), "Only support index keys that are a single column oid");
+        for (uint32_t col_idx = 0; col_idx < num_index_cols; col_idx++) {
+          const auto &col = key_schema_.GetColumn(col_idx);
+          auto index_col_oid = col.Oid();
+          const catalog::col_oid_t &table_col_oid = indexed_attributes[col_idx];
+          if (table_pr->IsNull(pr_map[table_col_oid])) {
+            index_pr->SetNull(index->GetKeyOidToOffsetMap().at(index_col_oid));
+          } else {
+            auto size = AttrSizeBytes(col.AttrSize());
+            std::memcpy(index_pr->AccessForceNotNull(index->GetKeyOidToOffsetMap().at(index_col_oid)),
+                        table_pr->AccessWithNullCheck(pr_map[table_col_oid]), size);
+          }
+        }
+
+        switch (type) {
+          case DataTable::VersionChainType::VISIBLE:
+            printf("Inserting into index");
+            index->Insert(txn_, *index_pr, slot);
+            break;
+          case DataTable::VersionChainType::INVISIBLE:
+          case DataTable::VersionChainType::PRE_UPDATE:
+            printf("Inserting into index");
+            index->Insert(txn_, *index_pr, slot);
+            printf("Deleting from index");
```

Comment:
At the end of the bulk inserts, what are the guarantees? (E.g., does the index contain valid entries for all txns that have committed at a particular timestamp, or does the index contain dirty entries which need to be rollbacked?)
Also, have you considered cases when transaction aborts and the version chains get rollback?

Reply:
These will be addressed in the design doc as well, but I'll respond here first

> At the end of the bulk inserts, what are the guarantees? 
The index contains valid entries for all transactions that may be in flight. The index cannot contain dirty entries which need to be rollbacked as we wait until any pre-existing modifying transactions are committed.

Any modifying transactions that want to perform modifications after we grab the lock on the table are forced to wait until the index is fully built after which point they will insert into the table using normal semantics (e.g. rolling back as necessary)

> Also, have you considered cases when transaction aborts and the version chains get rollback?
Yes - see the previous comment wrt waiting until in-flight transactions commit.